File: Solved/shiritori.py
n = int(input())
prev = input()
curr = ''
isFair = True
words = {prev}
for i in range(1,n): #starting from 1 because we already have the first word store in 'prev'
    curr = input()
    if prev[-1] != curr[0] or curr in words:
        if i%2 != 1: print('Player 1 Lost') #since we start with the second person's word
        else: print('Player 2 Lost')
        isFair = False
        break
    words.add(prev)
    prev = curr
    
if isFair: print('Fair Game')

File: Solved/Shiritori.java
import java.util.*;
class Shiritori{
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        String prev = sc.next();
        String curr = "";
        HashSet<String> words = new HashSet<>();
        words.add(prev);
        boolean isFair = true;
        for (int i = 1; i < n; i++) {
            curr = sc.next();
            if(prev.charAt(prev.length()-1) != curr.charAt(0) || words.contains(curr)){
                if (i % 2 != 1) System.out.println("Player 1 Lost");
                else System.out.println("Player 2 Lost");
                isFair = false;
                break;
            }
            words.add(curr);
            prev = curr;            
        }
        if (isFair) System.out.println("Fair Game");
        sc.close();
    }
}